### DISCUSSION TITLE
can I modify the flare code to create a new interface?

### DISCUSSION THREAD

**Discussion by BuzzLEquerre** [flareteam/flare-game#960] [General]
I started to create a flare mod with my graphics ect and I would like to modify the flare interface to have the game I imagine, I would like to know if this is possible?

**Comment by dorkster** [comment]
It's certainly possible. If you take a look at the [engine code](https://github.com/flareteam/flare-engine/tree/master/src), our in-game UI is broken up like so:

- The [Menu](https://github.com/flareteam/flare-engine/blob/master/src/Menu.cpp) class handles common functionality shared by all menus, such as the background image and alignment.
- Specific implementations for each menu are prefixed with "**Menu**", e.g. **MenuCharacter** and **MenuInventory**
- The [MenuManager](https://github.com/flareteam/flare-engine/blob/master/src/MenuManager.cpp) class handles the creation, rendering, and inter-operations of all the menus.

**Reply by BuzzLEquerre** [reply]
Okay, so I could change the images and the size of the interface buttons, that's great!
I had already managed to place my logo in the background but there is still the flare icon on the application. I don't know how to change it.
And I thank you for helping me I searched, you are the only forum that could guide me :)
![Screenshot_2024-08-09-20-43-19-235_com miui videoplayer](https://github.com/user-attachments/assets/66f0e3f4-4441-4776-a5e6-4668585a1ae0)

**Reply by dorkster** [reply]
The application icon can be altered by changing `images/logo/icon.png` in your mod.

I also forgot to mention that the title screen/config/new game/etc menus are prefixed with "**GameState**" instead of "**Menu**". e.g. **GameStateTitle** and **GameStateNew**.

Also, this is the official Flare forum, so you're in the right place :)

**Reply by BuzzLEquerre** [reply]
Okay, I'll test all that when I get back :)
I will come back to you without hesitation ! There I have a few hundred tiles to define  it's going to be long xd

**Reply by BuzzLEquerre** [reply]
Hello I would like to modify my interface, only before doing this I would like to change the game thumbnail as well as its name once launched, "flare" -> "kingdom's Stone". Only I can't find a file with the flare logo. I'd like to know if I can also modify the text that appears at the top right of the game "Flare 1.14 (Windows)" -> "Kingdom's Stone 0.1 (Windows)".
![Capture d'écran 2024-08-12 154127](https://github.com/user-attachments/assets/b937951a-fab1-44a5-a4ea-13c1c1da90e7)


As for the interface I'll find it myself but what I quoted I don't know, so if you have an idea to help me:)

**Reply by dorkster** [reply]
1. You can change the window title by setting the `window_title` property in `engine/misc.txt`
2. The text in the top right is supposed to be the *engine* version string, and is not meant to be changed. If you insist on changing it anyway, it's located [here](https://github.com/flareteam/flare-engine/blob/master/src/Version.h#L48) in the source code.
3. The method for setting the taskbar icon is platform-dependent. On Windows, it's done at compile-time, so we just [use the engine logo](https://github.com/flareteam/flare-engine/blob/master/src/Flare.rc#L18).

**Reply by BuzzLEquerre** [reply]
Okay so I have to modify the source code of the engine, only how do I access it? Should I start all over again?

**Reply by dorkster** [reply]
The engine source code is in this repo: https://github.com/flareteam/flare-engine

You can find our build instructions for Windows [here](https://github.com/flareteam/flare-engine/blob/master/INSTALL.engine.md#windows).

This will probably be daunting if you've never compiled something from C/C++ source before. It's also probably not worth your time if you're only interested in changing the engine branding. But if you're interested in learning how the engine works and possibly extending its functionality, I'd say give it a shot.

**Reply by BuzzLEquerre** [reply]
Wow, indeed it looks very difficult, especially since I have no basis and experience in this field. For the moment I'm going to just change the image and complete my mod.
Time and light, I could add it at the end, if by then I have learned to code.

And also one last question, all the engine files, how do I assemble them ?

**Reply by dorkster** [reply]
>And also one last question, all the engine files, how do I assemble them ?

Not exactly sure what you mean by "assemble" here. Can you clarify?

**Reply by BuzzLEquerre** [reply]
by that I mean, transform all the game code files into an app.exe

**Reply by dorkster** [reply]
Oh, I assumed you meant something else, because I already answered that. What you're asking is how to "compile", or "build", the engine. That involves following the [build instructions](https://github.com/flareteam/flare-engine/blob/master/INSTALL.engine.md#windows).

**Reply by BuzzLEquerre** [reply]
I'm sorry if I have trouble expressing myself, I don't know anything about computers and I'm not very good at speaking English. What I meant is that I can't use the code in the "INSTALL.ENGINE.md" folder, I saved my changes but nothing changed :
![Capture d'écran 2024-08-13 083936](https://github.com/user-attachments/assets/8b229faf-e23a-4fd5-ae4e-73d77ea24d64)

**Reply by dorkster** [reply]
Don't worry about it. English has plenty of opportunities to be confusing, even for native speakers.

The instructions for building on Windows are further down in `INSTALL.engine.md`. Look for the heading **MSYS2 / MinGW**.

1. Download and install [MSYS2](https://www.msys2.org/)
2. Open a MSYS2 shell; Look for **MSYS2 MinGW 64-bit** in your Start menu.
3. Pick one of the `pacman` commands based on your system. You'll most likely want 64-bit (the 2nd one)
  a. If you get an error, you may need to run `pacman -Syu` first
4. Once the packages are done updating & installing, you should be able to run the rest of the commands.

**Reply by BuzzLEquerre** [reply]
![Capture d'écran 2024-08-31 144719](https://github.com/user-attachments/assets/1fdfebb0-b074-4442-8fa8-8496a786a0ed)
help,
I understand nothing with the MSYS2 system, i can't place my editing flare engine files, I copyright the first command in the "~"app and I have a error mess, idk why so if you have idea for helping me, (the command is the first command in the build instruction for windows in 64bits)
![Uploading Capture d'écran 2024-08-31 140823.png…]()

**Reply by dorkster** [reply]
The first error is because you weren't able to change into the "flare-engine" directory. It looks like you skipped the git command, which downloads the repository into a directory with that name.

```
git clone https://github.com/flareteam/flare-engine.git
```

The next two errors are because you don't have the MinGW64 bin directory in your PATH environment variable. You can add it to you bash profile

```
echo "PATH=\$PATH:/mingw64/bin" >> ~/.bash_profile
source ~/.bash_profile
```

**Reply by BuzzLEquerre** [reply]
I used the cmake command to get my .exe file of my game but I can't create it I think I modified the flare.rc file in the game engine incorrectly, because the "flare-engine" file was created correctly : 
![Capture d'écran 2024-08-31 174636](https://github.com/user-attachments/assets/7c463a60-1797-482d-8e25-0a009e8efb23)

my flare.rc file : 

//
// The resource header (defines the icon).
//
#include "resource.h"

//
// Version information requires inclusion of either <windows.h> or <winver.h>
//
#include <WinVer.h>

// -------------------------------
//
// Icon
//

// Icon with lowest ID value placed first to ensure application icon
// remains consistent on all systems.
IDI_FLARE_APP_ICON        ICON                    "..\\distribution\\kingdoms_stone.png"

// -------------------------------
//
// Version information
//

VS_VERSION_INFO VERSIONINFO
  FILEVERSION    1,0,0,0
  PRODUCTVERSION 0,1,0,0
BEGIN
    BLOCK "StringFileInfo"
    BEGIN
        BLOCK "040904E4"
        BEGIN
            VALUE "Comments", "Kingdom's Stone\0"
            VALUE "FileDescription", "Medieval dark fantaisy RPG\0"
            VALUE "FileVersion", "1, 0, 0\0"
            VALUE "InternalName", "Kingdom's Stone\0"
            VALUE "OriginalFilename", "kingdoms_stone.exe\0"
            VALUE "ProductName", "Kingdom's Stone\0"
            VALUE "ProductVersion", "0, 1, 0\0"
        END
    END
    BLOCK "VarFileInfo"
    BEGIN
		VALUE "Translation", 0x000, 1200
    END
END

**Reply by dorkster** [reply]
You cloned the **flare-game** repository, which does not have the engine source code. You need to get back to the **flare-engine** directory in order to compile the source code. Try this first:

```
cd ../flare-engine
```

If that gives you an error, do this:

```
cd ../
git clone https://github.com/flareteam/flare-engine.git
cd flare-engine/
```

Then you can run cmake to create the MakeFile and mingw32-make to compile it:

```
cmake . -DCMAKE_BUILD_TYPE=Release -G "MinGW Makefiles"
mingw32-make
```

**Reply by BuzzLEquerre** [reply]
I added my mods as well as the modified files in "scr" ex "flare.rc" and there seems to be a problem : 
![Capture d'écran 2024-08-31 182940](https://github.com/user-attachments/assets/4c24d94c-5dfc-4465-8a17-052facbd4454)

**Reply by dorkster** [reply]
Looks like all you need to do now is copy kingdoms_stone.png to the distribution folder inside the flare-engine folder. Then just run mingw32-make again.

**Comment by BuzzLEquerre** [comment]
![kingdoms_stone png](https://github.com/user-attachments/assets/4dd42611-1b27-4f44-91fd-2db56dee25da)
I place "kingdoms_stone.png" in distribution file but nothing appen :/

![Capture d'écran 2024-08-31 184624](https://github.com/user-attachments/assets/33bbf495-fec3-40e1-b7b5-7e875d12f82a)

**Reply by dorkster** [reply]
Sorry, I wasn't thinking. I think there's 2 problems there:

1. You can only use *.ico files. You'll have to convert your PNG. I think GIMP should be able to export to .ico.
2. You need to escape the backslashes in the path to your icon. Escaping is done with the backslash character, so each `\` becomes `\\`. So once you convert your icon, this line:
```
IDI_FLARE_APP_ICON ICON "..\distribution\kingdoms_stone.png"
```
should become this:
```
IDI_FLARE_APP_ICON ICON "..\\distribution\\kingdoms_stone.ico"
```

**Reply by BuzzLEquerre** [reply]
i already have this line in my flare.rc file, and I convert my PNG in ico, but the problem is the same
![Capture d'écran 2024-08-31 195816](https://github.com/user-attachments/assets/e5cbd6d8-e558-4c6b-93b0-e6580ba08b16)

**Reply by dorkster** [reply]
Ok, not sure why that's not working now. What happens if you try this command?

```
file distribution/kingdoms_stone.ico
```

It should give some output similar to this (I ran it on Flare.ico instead):
```
distribution/Flare.ico: MS Windows icon resource - 6 icons, 256x256 with PNG image data, 256 x 256, 8-bit/color RGBA, non-interlaced, 32 bits/pixel, -128x-128, 32 bits/pixel
```

**Reply by BuzzLEquerre** [reply]
It's return :
$ file distribution/kingdoms_stone.ico
distribution/kingdoms_stone.ico: cannot open `distribution/kingdoms_stone.ico' (No such file or directory)
![Capture d'écran 2024-08-31 205716](https://github.com/user-attachments/assets/92e17f0d-1e5a-451d-bc82-a86196aa6253)
![Capture d'écran 2024-08-31 205941](https://github.com/user-attachments/assets/0092d02e-2c01-4758-a762-12c969c245b2)

**Reply by dorkster** [reply]
Hmm, what about the outputs of

```
pwd
```
and
```
ls -l distribution/
```

**Reply by BuzzLEquerre** [reply]
![Capture d'écran 2024-08-31 210311](https://github.com/user-attachments/assets/75bac9fd-0339-4428-80b5-2f7b98e83d7c)

**Reply by dorkster** [reply]
Ah! I see the problem. Run this:

```
mv distribution/kingdoms_stone.ico.ico distribution/kingdoms_stone.ico
```
Windows Explorer hides file extensions by default. I recommend you turn them on (this [Youtube video](https://www.youtube.com/watch?v=z5FBLAagPIc) shows how).

**Reply by BuzzLEquerre** [reply]
![Capture d'écran 2024-08-31 211034](https://github.com/user-attachments/assets/b30bfa42-6a50-4a4d-8ec5-7ee1bde500b8)
Yes ! I have the .exe file, but it's name is Flare and in the Flare.rc component I replace the name by "Kingdom's Stone" why I have Flare in .exe name ? And my game icon failed, it's strange..

**Reply by dorkster** [reply]
Our CMakeLists.txt file is set up to output the binary as flare.exe. You can change the instances of "flare" in that file and re-run the cmake command, but you might break the build. It might be easier to just rename the exe file to whatever you want.

Not sure why the icon isn't working. Maybe it needs a clean build?
```
mingw32-make clean
mingw32-make
```

I do question as to why you removed the copyright line from the resource file. I don't think the GPLv3 allows you to just strip out copyright notices like that, but I am not a lawyer. The proper way would be to preserve the original copyright and ADD your own copyright to it. Otherwise, it would be unfair to Clint, myself, and all the other developers that have worked on Flare over the years.

**Reply by BuzzLEquerre** [reply]
oh, that wasn't my intention, I didn't think leaving the line in the app was useful since there are the credits in the game and the copyright files in the game file. I'll put it back, I didn't know it was very important.

I restart the processus for create the new flare.exe, and i'm very sorry for the problem, on the contrary I respect your work enormously and I love the fact that you help people who know nothing about it like me :)
![Capture d'écran 2024-08-31 221701](https://github.com/user-attachments/assets/891b74a8-8a96-41fb-80f8-bed2abcd005e)
![Capture d'écran 2024-08-31 221931](https://github.com/user-attachments/assets/1dc47e12-e6fb-4d0b-8ef1-a6ff351e394f)

My god! I just had to rename flare.exe to Kingdom's Stone.exe! Thank you so much for helping me! Without you I would never have managed to have an app with my icon !
![Capture d'écran 2024-08-31 222813](https://github.com/user-attachments/assets/769d7c19-5567-425d-aa13-0e64e92e685f)

**Reply by dorkster** [reply]
Sorry for bringing up the copyright stuff at all. I didn't think you had any bad intent. We've just had cases in the past where people ignored the license, so sometimes it's best to clarify things up front.

Anyway, I guess Windows Explorer had cached the icon, so renaming the exe forced it to refresh. Maybe refreshing the folder view with F5 would have worked as well?

And feel free to report any bugs. The development version has several fixes since 1.14, but I may have broken some stuff as well :)

**Reply by dorkster** [reply]
Oh, and here's a tip if you make more changes to the engine: When you run mingw32-make, you can tell it to use multiple threads with the `-j` flag to speed things up. For example, my computer has a 16 core CPU, so I'll run:
```
mingw32-make -j16
```

**Reply by BuzzLEquerre** [reply]
I have not yet added the modifications to the inventory, but I have not found any bugs for the moment ^^ and yes I reassure you I want to do this completely legally I am not trying to steal the work of others :)

**Reply by BuzzLEquerre** [reply]
On my PC it says "intel core i5 10th gen" so I have 5 cores ?

**Reply by dorkster** [reply]
"i5" is just Intel's branding. There's several models that fall under that description, but I think the minimum is 4 physical cores / 8 threads. So you could probably use `-j8` without issue. The `-j` flag is to set the number of jobs for make to create. You can create more jobs than you have CPU threads, but it's generally inefficient if you go above your thread count.

You can find out more about your CPU by opening Task Manager (Ctrl+Shift+Esc), and then going to the Performance tab and selecting CPU. You should be able to find your model name there, as well as how many "virtual" processors you have.

**Reply by BuzzLEquerre** [reply]
I just noticed that there is a problem, when Kingdom's Stone is not in full screen at the top of the window it says "Flare" instead of Kingdom's Stone, and at the top right the game version is not "windows 0.1" as I had indicated in the "flare.rc" file.
Do you know where I can change this ?
![Capture d'écran 2024-09-01 094109](https://github.com/user-attachments/assets/09413318-f252-4d1b-aeee-37da96a3c21e)

**Reply by dorkster** [reply]
>when Kingdom's Stone is not in full screen at the top of the window it says "Flare" instead of Kingdom's Stone

Make sure you have `window_title=Kingdom's Stone` in your mod's `engine/misc.txt` file. That should be all you need to change that.

>at the top right the game version is not "windows 0.1" as I had indicated in the "flare.rc" file. Do you know where I can change this ?

The version information displayed in-game is specified in `src/Version.cpp` and `src/Version.h`. You already changed the `VersionInfo.NAME` in `src/Version.h`. The version number is set with `VersionInfo::ENGINE` in `src/Version.cpp`.

However, I **highly** recommend not changing the engine version number. Just makes it more difficult to compare your fork to the upstream version of Flare, and it'll be one more conflict you need to deal with if you want to merge a later upstream version of Flare back into your fork.

**Reply by BuzzLEquerre** [reply]
Right, ok I add "window_title=Kingdom's Stone" in the good file, thk you it's work :)
Oh, ok I see, so I have to modify the game version once i finish it !
